File: backend/game/kernel.py
"""Deterministic six-turn night kernel. No LLM, no database."""

from __future__ import annotations

import copy
import hashlib
import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any


def _load_scenario() -> dict[str, Any]:
    here = Path(__file__).resolve()
    candidates = (
        here.parents[2] / "src" / "features" / "game" / "one_night.json",
        here.parent / "data" / "one_night.json",
    )
    for path in candidates:
        if path.is_file():
            return json.loads(path.read_text(encoding="utf-8"))
    raise FileNotFoundError("one_night.json")


SCENARIO = _load_scenario()
TURNS = int(SCENARIO["turns"])
METER_BOUNDS = {
    name: (int(spec["min"]), int(spec["max"])) for name, spec in SCENARIO["meters"].items()
}


def _clamp_meter(name: str, value: int) -> int:
    lo, hi = METER_BOUNDS[name]
    return max(lo, min(hi, value))


@dataclass
class GameState:
    run_id: str
    seed: int
    revision: int
    turn: int
    meters: dict[str, int]
    resources: dict[str, int]
    location: str
    flags: dict[str, Any]
    known: list[str]
    promises: list[dict[str, Any]]
    fired_npc: list[str]
    log: list[dict[str, Any]]
    scene: dict[str, str]
    last_consequence: str
    ending: dict[str, Any] | None = None
    last_actions: list[dict[str, Any]] = field(default_factory=list)


def start_run(seed: int = 1, run_id: str | None = None) -> GameState:
    meters = {name: int(spec["start"]) for name, spec in SCENARIO["meters"].items()}
    resources = {name: int(spec["start"]) for name, spec in SCENARIO["resources"].items()}
    return GameState(
        run_id=run_id or f"night-{seed}",
        seed=seed,
        revision=0,
        turn=1,
        meters=meters,
        resources=resources,
        location=str(SCENARIO["location_start"]),
        flags=copy.deepcopy(SCENARIO["flags_start"]),
        known=[],
        promises=copy.deepcopy(SCENARIO["promises_start"]),
        fired_npc=[],
        log=[],
        scene=copy.deepcopy(SCENARIO["opening_scene"]),
        last_consequence="",
        ending=None,
        last_actions=[],
    )


def fingerprint(state: GameState) -> str:
    payload = {
        "seed": state.seed,
        "revision": state.revision,
        "turn": state.turn,
        "meters": state.meters,
        "resources": state.resources,
        "location": state.location,
        "flags": state.flags,
        "known": state.known,
        "promises": state.promises,
        "fired_npc": state.fired_npc,
        "log": state.log,
        "ending": state.ending,
    }
    blob = json.dumps(payload, sort_keys=True, default=str, ensure_ascii=False)
    return hashlib.sha256(blob.encode("utf-8")).hexdigest()


def _flag_value(state: GameState, key: str) -> Any:
    return state.flags.get(key)


def _promise(state: GameState, pid: str) -> dict[str, Any] | None:
    for item in state.promises:
        if item.get("id") == pid:
            return item
    return None


def _promise_open(state: GameState, pid: str) -> bool:
    found = _promise(state, pid)
    return bool(found) and not found.get("resolved")


def matches(when: dict[str, Any] | None, state: GameState) -> bool:
    if not when:
        return True
    flag_eq = when.get("flag_eq") or {}
    for key, expected in flag_eq.items():
        if _flag_value(state, key) != expected:
            return False
    flag_in = when.get("flag_in") or {}
    for key, options in flag_in.items():
        if _flag_value(state, key) not in options:
            return False
    resource_gte = when.get("resource_gte") or {}
    for key, minimum in resource_gte.items():
        if int(state.resources.get(key, 0)) < int(minimum):
            return False
    known = when.get("known") or []
    for fact in known:
        if fact not in state.known:
            return False
    if "location" in when and state.location != when["location"]:
        return False
    if "promise_open" in when and not _promise_open(state, when["promise_open"]):
        return False
    if "promise_missing" in when and _promise_open(state, when["promise_missing"]):
        return False
    if "promise_resolved" in when:
        found = _promise(state, when["promise_resolved"])
        if not found or not found.get("resolved"):
            return False
    meter_gte = when.get("meter_gte") or {}
    for key, minimum in meter_gte.items():
        if int(state.meters.get(key, 0)) < int(minimum):
            return False
    meter_lte = when.get("meter_lte") or {}
    for key, maximum in meter_lte.items():
        if int(state.meters.get(key, 0)) > int(maximum):
            return False
    return True


def _legal_defs(state: GameState) -> list[dict[str, Any]]:
    matched = [a for a in SCENARIO["actions"] if matches(a.get("when"), state)]
    if len(matched) <= 3:
        return matched
    without = [a for a in matched if a["id"] != "stall"]
    if len(without) >= 3:
        return without[:3]
    stall = [a for a in matched if a["id"] == "stall"]
    return (without + stall)[:3]


def _append_log(
    state: GameState,
    *,
    source: str,
    text: str,
    action_id: str | None = None,
    label: str = "",
    acted_on_turn: int | None = None,
) -> None:
    if not text:
        return
    state.log.append(
        {
            "source": source,
            "text": text,
            "action_id": action_id,
            "label": label,
            "turn": acted_on_turn if acted_on_turn is not None else state.turn,
        }
    )


def _apply_effects(
    state: GameState,
    effects: dict[str, Any] | None,
    *,
    source: str,
    action_id: str | None = None,
    label: str = "",
    acted_on_turn: int | None = None,
) -> None:
    if not effects:
        return
    for name, delta in (effects.get("meters") or {}).items():
        state.meters[name] = _clamp_meter(name, int(state.meters.get(name, 0)) + int(delta))
    for name, delta in (effects.get("resources") or {}).items():
        nxt = int(state.resources.get(name, 0)) + int(delta)
        state.resources[name] = max(0, nxt)
    for name, value in (effects.get("flags") or {}).items():
        state.flags[name] = value
    if "location" in effects:
        state.location = str(effects["location"])
    for fact in effects.get("known") or []:
        if fact not in state.known:
            state.known.append(fact)
    pid = effects.get("resolve_promise")
    if pid:
        found = _promise(state, pid)
        if found:
            found["resolved"] = True
    delay = effects.get("delay_promise")
    if delay:
        found = _promise(state, delay["id"])
        if found and not found.get("resolved"):
            found["due_turn"] = int(found.get("due_turn") or 0) + int(delay.get("by") or 0)
    created = effects.get("create_promise")
    if created and not _promise(state, created["id"]):
        item = copy.deepcopy(created)
        item.setdefault("resolved", False)
        item.setdefault("triggered", False)
        if action_id and not item.get("source_action_id"):
            item["source_action_id"] = action_id
        state.promises.append(item)
    scene = effects.get("scene")
    if scene:
        state.scene = copy.deepcopy(scene)
    _append_log(
        state,
        source=source,
        text=str(effects.get("log") or ""),
        action_id=action_id,
        label=label,
        acted_on_turn=acted_on_turn,
    )


def _apply_block(
    state: GameState,
    spec: dict[str, Any],
    *,
    source: str,
    action_id: str | None = None,
    acted_on_turn: int | None = None,
) -> None:
    label = str(spec.get("label") or "")
    _apply_effects(
        state,
        spec.get("effects"),
        source=source,
        action_id=action_id,
        label=label,
        acted_on_turn=acted_on_turn,
    )
    for extra in spec.get("effects_if") or []:
        if matches(extra.get("when"), state):
            _apply_effects(
                state,
                extra,
                source=source,
                action_id=action_id,
                label=label,
                acted_on_turn=acted_on_turn,
            )


def _run_npcs(state: GameState) -> None:
    for step in SCENARIO["npc_steps"]:
        sid = str(step["id"])
        if step.get("once") and sid in state.fired_npc:
            continue
        if not matches(step.get("when"), state):
            continue
        _apply_effects(state, step.get("effects"), source="npc")
        if step.get("once"):
            state.fired_npc.append(sid)


def _run_promises(state: GameState) -> None:
    for item in state.promises:
        if item.get("resolved") or item.get("triggered"):
            continue
        if state.turn < int(item.get("due_turn") or 0):
            continue
        item["triggered"] = True
        _apply_effects(state, item.get("overdue"), source="promise")
        _append_log(state, source="promise", text=str(item.get("overdue_log") or ""))


def _ending_causes(state: GameState) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    seen: set[tuple[Any, Any]] = set()
    for entry in state.log:
        if entry.get("source") != "player" or not entry.get("action_id"):
            continue
        key = (entry.get("turn"), entry.get("action_id"))
        if key in seen:
            continue
        seen.add(key)
        out.append(
            {
                "turn": entry.get("turn"),
                "action_id": entry.get("action_id"),
                "label": entry.get("label") or "",
            }
        )
    return out


def _evaluate_ending(state: GameState) -> dict[str, Any] | None:
    for spec in SCENARIO["endings"]:
        if not spec.get("hard"):
            continue
        if matches(spec.get("when"), state):
            return {
                "id": spec["id"],
                "title": spec["title"],
                "body": spec["body"],
                "causes": _ending_causes(state),
            }
    if state.turn <= TURNS:
        return None
    for spec in SCENARIO["endings"]:
        if spec.get("hard"):
            continue
        if matches(spec.get("when"), state):
            return {
                "id": spec["id"],
                "title": spec["title"],
                "body": spec["body"],
                "causes": _ending_causes(state),
            }
    return None


def apply_action(state: GameState, action_id: str) -> tuple[GameState, str | None]:
    nxt = copy.deepcopy(state)
    if nxt.ending:
        return state, "already_ended"
    legal = _legal_defs(nxt)
    spec = next((a for a in legal if a["id"] == action_id), None)
    if spec is None:
        return state, "illegal_action"
    acted_on = nxt.turn
    _apply_block(nxt, spec, source="player", action_id=action_id, acted_on_turn=acted_on)
    nxt.turn += 1
    nxt.revision += 1
    _run_npcs(nxt)
    _run_promises(nxt)
    nxt.ending = _evaluate_ending(nxt)
    nxt.last_actions = [
        {"id": a["id"], "label": a["label"], "cost_text": a["cost_text"]} for a in _legal_defs(nxt)
    ]
    bits = [e.get("text") for e in nxt.log if e.get("turn") == acted_on]
    nxt.last_consequence = " ".join(t for t in bits if t)
    return nxt, None


def _visible_history(state: GameState) -> list[dict[str, Any]]:
    specs = {a["id"]: a for a in SCENARIO["actions"]}
    out: list[dict[str, Any]] = []
    seen_player: set[tuple[Any, Any]] = set()
    for entry in state.log:
        source = entry.get("source")
        if source not in ("player", "npc", "promise"):
            continue
        text = str(entry.get("text") or "").strip()
        if not text:
            continue
        if source == "player":
            key = (entry.get("turn"), entry.get("action_id"))
            if key in seen_player:
                for item in reversed(out):
                    if (
                        item.get("source") == "player"
                        and item.get("turn") == key[0]
                        and item.get("action_id") == key[1]
                    ):
                        item["text"] = f"{item['text']} {text}".strip()
                        break
                continue
            seen_player.add(key)
            spec = specs.get(str(entry.get("action_id") or "")) or {}
            out.append(
                {
                    "turn": entry.get("turn"),
                    "source": "player",
                    "text": text,
                    "action_id": entry.get("action_id"),
                    "label": entry.get("label") or spec.get("label") or "",
                    "cost_text": spec.get("cost_text") or "",
                }
            )
            continue
        out.append(
            {
                "turn": entry.get("turn"),
                "source": source,
                "text": text,
                "action_id": entry.get("action_id"),
                "label": entry.get("label") or "",
                "cost_text": "",
            }
        )
    return out


def player_view(state: GameState) -> dict[str, Any]:
    labels = SCENARIO.get("location_labels") or {}
    legal = [
        {"id": a["id"], "label": a["label"], "cost_text": a["cost_text"]}
        for a in (state.last_actions or _legal_defs(state))
    ]
    if not state.last_actions and not state.ending:
        legal = [
            {"id": a["id"], "label": a["label"], "cost_text": a["cost_text"]}
            for a in _legal_defs(state)
        ]
    promises = []
    for item in state.promises:
        if item.get("resolved"):
            continue
        promises.append(
            {
                "id": item["id"],
                "label": item.get("label") or "",
                "due_turn": item.get("due_turn"),
                "source_action_id": item.get("source_action_id"),
            }
        )
    history = _visible_history(state)
    return {
        "run_id": state.run_id,
        "revision": state.revision,
        "turn": min(state.turn, TURNS) if not state.ending else TURNS,
        "turns_max": TURNS,
        "location": state.location,
        "location_label": labels.get(state.location, state.location),
        "player": SCENARIO["player"],
        "objective": SCENARIO["objective"],
        "meters": dict(state.meters),
        "resources": dict(state.resources),
        "promises": promises,
        "known": list(state.known),
        "scene": dict(state.scene),
        "last_consequence": state.last_consequence,
        "legal_actions": [] if state.ending else legal,
        "ending": copy.deepcopy(state.ending),
        "history": history,
        "visibility": "player",
        "performance": None,
    }


def state_to_dict(state: GameState) -> dict[str, Any]:
    return {
        "run_id": state.run_id,
        "seed": state.seed,
        "revision": state.revision,
        "turn": state.turn,
        "meters": copy.deepcopy(state.meters),
        "resources": copy.deepcopy(state.resources),
        "location": state.location,
        "flags": copy.deepcopy(state.flags),
        "known": list(state.known),
        "promises": copy.deepcopy(state.promises),
        "fired_npc": list(state.fired_npc),
        "log": copy.deepcopy(state.log),
        "scene": copy.deepcopy(state.scene),
        "last_consequence": state.last_consequence,
        "ending": copy.deepcopy(state.ending),
        "last_actions": copy.deepcopy(state.last_actions),
    }


def state_from_dict(data: dict[str, Any]) -> GameState:
    payload = copy.deepcopy(data)
    return GameState(
        run_id=str(payload["run_id"]),
        seed=int(payload["seed"]),
        revision=int(payload["revision"]),
        turn=int(payload["turn"]),
        meters=dict(payload["meters"]),
        resources=dict(payload["resources"]),
        location=str(payload["location"]),
        flags=dict(payload["flags"]),
        known=list(payload.get("known") or []),
        promises=list(payload.get("promises") or []),
        fired_npc=list(payload.get("fired_npc") or []),
        log=list(payload.get("log") or []),
        scene=dict(payload.get("scene") or {"speaker": "", "body": ""}),
        last_consequence=str(payload.get("last_consequence") or ""),
        ending=payload.get("ending"),
        last_actions=list(payload.get("last_actions") or []),
    )

File: src/features/game/kernel.ts
import scenarioJson from './one_night.json' with { type: 'json' }
import type { GameActionChoice, GameEnding, GameHistoryItem, PlayerView } from './contracts.ts'

type Scenario = typeof scenarioJson
type ActionSpec = Scenario['actions'][number]
type PromiseRec = {
  id: string
  owed_to: string
  due_turn: number
  source_action_id: string | null
  label: string
  overdue_log: string
  overdue?: Record<string, unknown>
  resolved?: boolean
  triggered?: boolean
}

const SCENARIO = scenarioJson
const TURNS = SCENARIO.turns
const METER_BOUNDS = Object.fromEntries(
  Object.entries(SCENARIO.meters).map(([name, spec]) => [name, [spec.min, spec.max] as const]),
)

export type GameState = {
  run_id: string
  seed: number
  revision: number
  turn: number
  meters: Record<string, number>
  resources: Record<string, number>
  location: string
  flags: Record<string, unknown>
  known: string[]
  promises: PromiseRec[]
  fired_npc: string[]
  log: Array<{
    source: string
    text: string
    action_id?: string | null
    label?: string
    turn: number
  }>
  scene: { speaker: string; body: string }
  last_consequence: string
  ending: GameEnding | null
  last_actions: GameActionChoice[]
}

function clone<T>(value: T): T {
  return structuredClone(value)
}

function clampMeter(name: string, value: number): number {
  const bounds = METER_BOUNDS[name]
  if (!bounds) return value
  return Math.max(bounds[0], Math.min(bounds[1], value))
}

function promiseOf(state: GameState, pid: string): PromiseRec | undefined {
  return state.promises.find((item) => item.id === pid)
}

function promiseOpen(state: GameState, pid: string): boolean {
  const found = promiseOf(state, pid)
  return Boolean(found) && !found?.resolved
}

function asRecord(value: unknown): Record<string, unknown> {
  return value && typeof value === 'object' && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : {}
}

function matches(when: unknown, state: GameState): boolean {
  const clause = asRecord(when)
  if (Object.keys(clause).length === 0) return true
  for (const [key, expected] of Object.entries(asRecord(clause.flag_eq))) {
    if (state.flags[key] !== expected) return false
  }
  for (const [key, options] of Object.entries(asRecord(clause.flag_in))) {
    if (!Array.isArray(options) || !options.includes(state.flags[key])) return false
  }
  for (const [key, minimum] of Object.entries(asRecord(clause.resource_gte))) {
    if ((state.resources[key] ?? 0) < Number(minimum)) return false
  }
  const known = clause.known
  if (Array.isArray(known)) {
    for (const fact of known) {
      if (typeof fact === 'string' && !state.known.includes(fact)) return false
    }
  }
  if ('location' in clause && state.location !== clause.location) return false
  if ('promise_open' in clause && typeof clause.promise_open === 'string' && !promiseOpen(state, clause.promise_open)) {
    return false
  }
  if (
    'promise_missing' in clause &&
    typeof clause.promise_missing === 'string' &&
    promiseOpen(state, clause.promise_missing)
  ) {
    return false
  }
  if ('promise_resolved' in clause && typeof clause.promise_resolved === 'string') {
    const found = promiseOf(state, clause.promise_resolved)
    if (!found?.resolved) return false
  }
  for (const [key, minimum] of Object.entries(asRecord(clause.meter_gte))) {
    if ((state.meters[key] ?? 0) < Number(minimum)) return false
  }
  for (const [key, maximum] of Object.entries(asRecord(clause.meter_lte))) {
    if ((state.meters[key] ?? 0) > Number(maximum)) return false
  }
  return true
}

function legalDefs(state: GameState): ActionSpec[] {
  const matched = SCENARIO.actions.filter((action) => matches(action.when, state))
  if (matched.length <= 3) return matched
  const without = matched.filter((action) => action.id !== 'stall')
  if (without.length >= 3) return without.slice(0, 3)
  const stall = matched.filter((action) => action.id === 'stall')
  return [...without, ...stall].slice(0, 3)
}

function appendLog(
  state: GameState,
  opts: {
    source: string
    text: string
    actionId?: string | null
    label?: string
    actedOnTurn?: number
  },
): void {
  if (!opts.text) return
  state.log.push({
    source: opts.source,
    text: opts.text,
    action_id: opts.actionId,
    label: opts.label ?? '',
    turn: opts.actedOnTurn ?? state.turn,
  })
}

function applyEffects(
  state: GameState,
  effects: unknown,
  opts: { source: string; actionId?: string | null; label?: string; actedOnTurn?: number },
): void {
  const body = asRecord(effects)
  if (Object.keys(body).length === 0) return
  for (const [name, delta] of Object.entries(asRecord(body.meters))) {
    state.meters[name] = clampMeter(name, (state.meters[name] ?? 0) + Number(delta))
  }
  for (const [name, delta] of Object.entries(asRecord(body.resources))) {
    state.resources[name] = Math.max(0, (state.resources[name] ?? 0) + Number(delta))
  }
  for (const [name, value] of Object.entries(asRecord(body.flags))) {
    state.flags[name] = value
  }
  if ('location' in body) state.location = String(body.location)
  const known = body.known
  if (Array.isArray(known)) {
    for (const fact of known) {
      if (typeof fact === 'string' && !state.known.includes(fact)) state.known.push(fact)
    }
  }
  if (typeof body.resolve_promise === 'string') {
    const found = promiseOf(state, body.resolve_promise)
    if (found) found.resolved = true
  }
  const delay = asRecord(body.delay_promise)
  if (typeof delay.id === 'string') {
    const found = promiseOf(state, delay.id)
    if (found && !found.resolved) {
      found.due_turn = Number(found.due_turn ?? 0) + Number(delay.by ?? 0)
    }
  }
  const created = asRecord(body.create_promise)
  if (typeof created.id === 'string' && !promiseOf(state, created.id)) {
    const item = clone(created) as PromiseRec
    item.resolved = false
    item.triggered = false
    if (opts.actionId && !item.source_action_id) item.source_action_id = opts.actionId
    state.promises.push(item)
  }
  const scene = asRecord(body.scene)
  if (typeof scene.body === 'string') {
    state.scene = { speaker: String(scene.speaker ?? ''), body: scene.body }
  }
  appendLog(state, {
    source: opts.source,
    text: String(body.log ?? ''),
    actionId: opts.actionId,
    label: opts.label,
    actedOnTurn: opts.actedOnTurn,
  })
}

function applyBlock(
  state: GameState,
  spec: { label?: string; effects?: unknown; effects_if?: unknown },
  opts: { source: string; actionId?: string | null; actedOnTurn?: number },
): void {
  const label = spec.label ?? ''
  applyEffects(state, spec.effects, { ...opts, label })
  const extras = Array.isArray(spec.effects_if) ? spec.effects_if : []
  for (const extra of extras) {
    const row = asRecord(extra)
    if (matches(row.when, state)) applyEffects(state, extra, { ...opts, label })
  }
}

function runNpcs(state: GameState): void {
  for (const step of SCENARIO.npc_steps) {
    if (step.once && state.fired_npc.includes(step.id)) continue
    if (!matches(step.when, state)) continue
    applyEffects(state, step.effects, { source: 'npc' })
    if (step.once) state.fired_npc.push(step.id)
  }
}

function runPromises(state: GameState): void {
  for (const item of state.promises) {
    if (item.resolved || item.triggered) continue
    if (state.turn < Number(item.due_turn ?? 0)) continue
    item.triggered = true
    applyEffects(state, 'overdue' in item ? item.overdue : undefined, { source: 'promise' })
    appendLog(state, { source: 'promise', text: String(item.overdue_log ?? '') })
  }
}

function endingCauses(state: GameState): GameEnding['causes'] {
  const out: GameEnding['causes'] = []
  const seen = new Set<string>()
  for (const entry of state.log) {
    if (entry.source !== 'player' || !entry.action_id) continue
    const key = `${entry.turn}:${entry.action_id}`
    if (seen.has(key)) continue
    seen.add(key)
    out.push({
      turn: entry.turn,
      action_id: entry.action_id,
      label: entry.label ?? '',
    })
  }
  return out
}

function evaluateEnding(state: GameState): GameEnding | null {
  for (const spec of SCENARIO.endings) {
    if (!spec.hard || !matches(spec.when, state)) continue
    return { id: spec.id, title: spec.title, body: spec.body, causes: endingCauses(state) }
  }
  if (state.turn <= TURNS) return null
  for (const spec of SCENARIO.endings) {
    if (spec.hard) continue
    if (!matches(spec.when, state)) continue
    return { id: spec.id, title: spec.title, body: spec.body, causes: endingCauses(state) }
  }
  return null
}

function toChoices(defs: ActionSpec[]): GameActionChoice[] {
  return defs.map((action) => ({ id: action.id, label: action.label, cost_text: action.cost_text }))
}

function visibleHistory(state: GameState): GameHistoryItem[] {
  const specs = Object.fromEntries(SCENARIO.actions.map((action) => [action.id, action]))
  const out: GameHistoryItem[] = []
  const seen = new Set<string>()
  for (const entry of state.log) {
    if (entry.source !== 'player' && entry.source !== 'npc' && entry.source !== 'promise') continue
    const text = entry.text.trim()
    if (!text) continue
    if (entry.source === 'player') {
      const key = `${entry.turn}:${entry.action_id ?? ''}`
      if (seen.has(key)) {
        const match = [...out].reverse().find(
          (item) => item.source === 'player' && item.turn === entry.turn && item.action_id === entry.action_id,
        )
        if (match) match.text = `${match.text} ${text}`.trim()
        continue
      }
      seen.add(key)
      const spec = entry.action_id ? specs[entry.action_id] : undefined
      out.push({
        turn: entry.turn,
        source: 'player',
        text,
        action_id: entry.action_id,
        label: entry.label || spec?.label || '',
        cost_text: spec?.cost_text ?? '',
      })
      continue
    }
    out.push({
      turn: entry.turn,
      source: entry.source,
      text,
      action_id: entry.action_id,
      label: entry.label ?? '',
      cost_text: '',
    })
  }
  return out
}

export function startRun(seed = 1): GameState {
  const meters = Object.fromEntries(
    Object.entries(SCENARIO.meters).map(([name, spec]) => [name, spec.start]),
  )
  const resources = Object.fromEntries(
    Object.entries(SCENARIO.resources).map(([name, spec]) => [name, spec.start]),
  )
  return {
    run_id: `night-${seed}`,
    seed,
    revision: 0,
    turn: 1,
    meters,
    resources,
    location: SCENARIO.location_start,
    flags: clone(SCENARIO.flags_start),
    known: [],
    promises: clone(SCENARIO.promises_start),
    fired_npc: [],
    log: [],
    scene: clone(SCENARIO.opening_scene),
    last_consequence: '',
    ending: null,
    last_actions: [],
  }
}

export function fingerprint(state: GameState): string {
  return JSON.stringify({
    seed: state.seed,
    revision: state.revision,
    turn: state.turn,
    meters: state.meters,
    resources: state.resources,
    location: state.location,
    flags: state.flags,
    known: state.known,
    promises: state.promises,
    fired_npc: state.fired_npc,
    log: state.log,
    ending: state.ending,
  })
}

export function applyAction(state: GameState, actionId: string): [GameState, string | null] {
  const next = clone(state)
  if (next.ending) return [state, 'already_ended']
  const spec = legalDefs(next).find((action) => action.id === actionId)
  if (!spec) return [state, 'illegal_action']
  const actedOn = next.turn
  applyBlock(next, spec, { source: 'player', actionId, actedOnTurn: actedOn })
  next.turn += 1
  next.revision += 1
  runNpcs(next)
  runPromises(next)
  next.ending = evaluateEnding(next)
  next.last_actions = toChoices(legalDefs(next))
  next.last_consequence = next.log
    .filter((entry) => entry.turn === actedOn)
    .map((entry) => entry.text)
    .filter(Boolean)
    .join(' ')
  return [next, null]
}

export function playerView(state: GameState): PlayerView {
  const labels = asRecord(SCENARIO.location_labels)
  const legal =
    state.last_actions.length > 0
      ? state.last_actions
      : toChoices(legalDefs(state))
  const promises = state.promises
    .filter((item) => !item.resolved)
    .map((item) => ({
      id: item.id,
      label: item.label,
      due_turn: item.due_turn ?? null,
      source_action_id: item.source_action_id ?? null,
    }))
  return {
    run_id: state.run_id,
    revision: state.revision,
    turn: state.ending ? TURNS : Math.min(state.turn, TURNS),
    turns_max: TURNS,
    location: state.location,
    location_label: String(labels[state.location] ?? state.location),
    player: SCENARIO.player,
    objective: SCENARIO.objective,
    meters: { ...state.meters },
    resources: { ...state.resources },
    promises,
    known: [...state.known],
    scene: { ...state.scene },
    last_consequence: state.last_consequence,
    legal_actions: state.ending ? [] : legal,
    ending: state.ending ? clone(state.ending) : null,
    history: visibleHistory(state),
  }
}
